### PR TITLE
CP-18251: Remove host ref in pool_update.detach and pool_update.attach

### PR DESCRIPTION
**Pull request by ghost**
1. Remove host ref in detach/attach
2. Adapt introduce to attach_helper/detach_helper
3. Remove fingerprint pins for pool_update

Signed-off-by: Cheng Zhang cheng.zhang@citrix.com
Signed-off-by: Phus Lu phus.lu@citrix.com

### CODE REVIEW

File: ocaml/xapi/message_forwarding.ml [hunk 1]
```diff
@@ -2671,23 +2673,26 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
...
-		let attach ~__context ~self ~host =
-			info "Pool_update.attach: pool update = '%s' host = '%s'" (pool_update_uuid ~__context self) (host_uuid ~__context host);
-			let local_fn = Local.Pool_update.attach ~self ~host in
-			do_op_on ~local_fn ~__context ~host
-				(fun session_id rpc -> Client.Pool_update.attach rpc session_id self host)
+		let attach ~__context ~self =
+			info "Pool_update.attach: pool update = '%s'" (pool_update_uuid ~__context self);
+			let local_fn = Local.Pool_update.attach ~self in
+			let vdi = Db.Pool_update.get_vdi ~__context ~self in
+			do_op_on ~local_fn ~__context ~host:(Xapi_pool_update.get_candidate_host ~__context ~vdi)
```
Comment (jonludlam) [inline ocaml/xapi/message_forwarding.ml]: Would using `forward_vdi_op` be useful here? I'm not sure the extra complexity of `get_candidate_host` is worth it.

File: ocaml/xapi/xapi_pool_update.ml [hunk 2]
```diff
@@ -120,47 +169,52 @@ let with_api_errors f x =
...
+  ^(Printf.sprintf ("[%s]\nname=%s\nbaseurl=%s\n") name_label name_label url)
+  ^(if signed then Printf.sprintf ("gpgkey=file:///etc/pki/rpm-gpg/%s") key else "")
+
+let attach_helper ~__context ~uuid ~vdi =
+  Mutex.execute updates_to_attach_count_tbl_mutex 
+  ( fun () ->
+    let mount_point_parent_dir = String.concat "/" [Xapi_globs.host_update_dir; uuid] in
+    let mount_point = Filename.concat mount_point_parent_dir "vdi" in
+    let host = Helpers.get_localhost ~__context in
+    let ip = try Db.Host.get_address ~__context ~self:host with _ -> "127.0.0.1" in
```
Comment (jonludlam) [inline ocaml/xapi/xapi_pool_update.ml]: I don't think you need the `try...with` here

File: ocaml/xapi/xapi_pool_update.ml [hunk 3]
```diff
@@ -120,47 +169,52 @@ let with_api_errors f x =
...
+    let url = "http://" ^ ip ^ Constants.get_pool_update_download_uri ^ uuid ^ "/vdi" in
+    let count = try Hashtbl.find updates_to_attach_count_tbl uuid with _ -> 0 in
+    if count = 0 then begin
+      debug "pool_update.attach_helper %s to %s" uuid mount_point;
+      if (try Sys.is_directory mount_point with _ -> false) then detach_helper ~__context ~uuid ~vdi;
+      let output, _ = Forkhelpers.execute_command_get_output "/bin/mkdir" ["-p"; mount_point] in
+      debug "pool_update.attach_helper Mountpoint created (output=%s)" output;
+      if vdi_visible ~__context ~vdi ~host then begin
```
Comment (jonludlam) [inline ocaml/xapi/xapi_pool_update.ml]: I think this is unnecessary. If the VDI isn't visible, message_forwarding has gone wrong.

File: ocaml/xapi/xapi_pool_update.ml [hunk 4]
```diff
@@ -182,17 +236,21 @@ let guidance_from_string = function
...
+let extract_update_info ~__context ~vdi ~verify =
+  let vdi_uuid = Db.VDI.get_uuid ~__context ~self:vdi in
   finally
     (fun () ->
-      let update_path = attach ~__context ~self ~host in
+      let url = attach_helper ~__context ~uuid:vdi_uuid ~vdi in
       let update_path =
-        if String.startswith "file://" update_path then
-          String.sub_to_end update_path (String.length "file://")
+        if String.startswith "http://" url then
```
Comment (jonludlam) [inline ocaml/xapi/xapi_pool_update.ml]: This makes no sense. The URL has just been constructed above from the uuid you've passed in, but you're extracting it again here?

File: ocaml/xapi/message_forwarding.ml (hunk 1 above)
Comment (cheng-z) [reply ocaml/xapi/message_forwarding.ml]: get_candidate_host will always return master if share SR.
Not sure forward_vdi_op have the same behavior? Simply read the code seems not guaranteed.

Comment (jonludlam) [reply ocaml/xapi/message_forwarding.ml]: Yes: https://github.com/xapi-project/xen-api/blob/069236fe70fa03d3938419fbfc49de9c5f113e69/ocaml/xapi/message_forwarding.ml#L185-L191

**Comment by cheng-z** [comment]
Hi @jonludlam 
Thanks a lot for your comments.
Please help to have a look at the latest code change.

**Comment by jonludlam** [comment]
I think we can still improve this by removing the need for `get_attached_host` - if we forward the pool_clean operation in the same way as the attach/detach I think we won't need it at all (and we can also drop the API call `Pool_update.clean` as well).

However, in the interests of unblocking GUI/testing on this, I'm going to merge this PR, although I would like to see a PR with the improvements I've just suggested soon!

**Pull request merged**